File: app/src/main/java/com/webkul/mobikul/odoo/fragment/ShippingMethodFragment.java
package com.webkul.mobikul.odoo.fragment;

import androidx.databinding.DataBindingUtil;
import android.os.Bundle;
import android.os.Parcelable;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioButton;

import com.webkul.mobikul.odoo.R;
import com.webkul.mobikul.odoo.databinding.FragmentShippingMethodBinding;
import com.webkul.mobikul.odoo.handler.checkout.ShippingMethodFragmentHandler;
import com.webkul.mobikul.odoo.model.checkout.ActiveShippingMethod;

import java.util.ArrayList;
import java.util.List;

import static com.webkul.mobikul.odoo.constant.ApplicationConstant.TAG_EXTRA_INFO;
import static com.webkul.mobikul.odoo.constant.BundleConstant.BUNDLE_KEY_ACTIVE_SHIPPING_DATA;


/**

 * Webkul Software.

 * @package Mobikul App

 * @Category Mobikul

 * @author Webkul <support@webkul.com>

 * @Copyright (c) Webkul Software Private Limited (https://webkul.com)

 * @license https://store.webkul.com/license.html ASL Licence

 * @link https://store.webkul.com/license.html

 */

public class ShippingMethodFragment extends BaseFragment {

    public FragmentShippingMethodBinding mBinding;

    public static ShippingMethodFragment newInstance(List<ActiveShippingMethod> shippingMethods) {
        Bundle args = new Bundle();
        args.putParcelableArrayList(BUNDLE_KEY_ACTIVE_SHIPPING_DATA, (ArrayList<? extends Parcelable>) shippingMethods);

        ShippingMethodFragment fragment = new ShippingMethodFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_shipping_method, container, false);
        return mBinding.getRoot();
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        ArrayList<ActiveShippingMethod> shippingMethodList = getArguments().getParcelableArrayList(BUNDLE_KEY_ACTIVE_SHIPPING_DATA);
        if (shippingMethodList != null) {
            for (ActiveShippingMethod shippingMethod : shippingMethodList) {
                RadioButton eachShippingMethodRb = new RadioButton(getContext());
//                if(Double.parseDouble(shippingMethod.getPrice())>0)
//                {
                eachShippingMethodRb.setText(String.format("%s (%s)", shippingMethod.getName(), shippingMethod.getPrice()));
//                }
//                else
//                {
//                    eachShippingMethodRb.setText(shippingMethod.getName());
//                }

                eachShippingMethodRb.setTag(shippingMethod.getId());
                mBinding.shippingMethodRg.addView(eachShippingMethodRb);
                if (!shippingMethod.getDescription().isEmpty()) {
                    AppCompatTextView extraInfoCompatTextView = new AppCompatTextView(getContext());
                    extraInfoCompatTextView.setText(shippingMethod.getDescription());
                    extraInfoCompatTextView.setTag(shippingMethod.getId() + TAG_EXTRA_INFO);
                    extraInfoCompatTextView.setVisibility(View.GONE);
                    mBinding.shippingMethodRg.addView(extraInfoCompatTextView);
                }
            }

            mBinding.shippingMethodRg.setOnCheckedChangeListener((group, checkedId) -> {
                String selectedShippingMethodId = group.findViewById(checkedId).getTag().toString();
                String selectedShippingMethodDescriptionTag = selectedShippingMethodId + TAG_EXTRA_INFO;
                for (ActiveShippingMethod shippingMethod : shippingMethodList) {
                    String eachViewTag = shippingMethod.getId() + TAG_EXTRA_INFO;
                    if (group.findViewWithTag(eachViewTag) != null) {
                        if (eachViewTag.equals(selectedShippingMethodDescriptionTag)) {
                            group.findViewWithTag(eachViewTag).setVisibility(View.VISIBLE);
                        } else {
                            group.findViewWithTag(eachViewTag).setVisibility(View.GONE);
                        }
                    }
                }
            });

            mBinding.setHandler(new ShippingMethodFragmentHandler(getContext()));

        }
    }
}

File: app/src/main/java/com/webkul/mobikul/odoo/fragment/ShippingMethodFragment.kt
package com.webkul.mobikul.odoo.fragment

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContract
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.Fragment
import cn.pedant.SweetAlert.SweetAlertDialog
import com.google.android.material.snackbar.Snackbar
import com.webkul.mobikul.odoo.R
import com.webkul.mobikul.odoo.activity.FragmentContainerActivity
import com.webkul.mobikul.odoo.activity.SignInSignUpActivity
import com.webkul.mobikul.odoo.adapter.checkout.ShippingMethodAdapter
import com.webkul.mobikul.odoo.adapter.checkout.ShippingMethodAdapter.ShippingMethodSelectionListener
import com.webkul.mobikul.odoo.connection.ApiConnection
import com.webkul.mobikul.odoo.connection.CustomObserver
import com.webkul.mobikul.odoo.constant.BundleConstant
import com.webkul.mobikul.odoo.databinding.FragmentNewShippingMethodBinding
import com.webkul.mobikul.odoo.helper.AlertDialogHelper
import com.webkul.mobikul.odoo.helper.AppSharedPref
import com.webkul.mobikul.odoo.helper.SnackbarHelper
import com.webkul.mobikul.odoo.model.BaseResponse
import com.webkul.mobikul.odoo.model.checkout.*
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers

class ShippingMethodFragment : Fragment(), ShippingMethodSelectionListener {

    lateinit var binding: FragmentNewShippingMethodBinding
    private val TAG_EXTRA_INFO = 1
    private var shippingMethod: ActiveShippingMethod? = null
    private var isShippingMethodSelected = false
    private var shippingMethodSelectionListener: ShippingMethodSelectionListener? = null
    private lateinit var dialog: SweetAlertDialog
    private val TAG = "ShippingMethodFragment"
    private var orderId: Int = 0


    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DataBindingUtil.inflate(
            inflater,
            R.layout.fragment_new_shipping_method,
            container,
            false
        )
        shippingMethodSelectionListener = this
        getArgument()
        setupDialog()
        fetchShippingMethods()
        setListeners()
        return binding.root
    }

    private fun setupDialog() {
        dialog = AlertDialogHelper.getAlertDialog(
            requireContext(), SweetAlertDialog.PROGRESS_TYPE,
            getString(R.string.shipping_options_loading_text), "", false, false
        )
        (requireActivity() as FragmentContainerActivity).setToolbarText(getString(R.string.shipping_method))
    }

    private fun getArgument() {
        orderId = arguments?.get(BundleConstant.BUNDLE_KEY_ORDER_ID) as Int
    }

    private fun fetchShippingMethods() {
        ApiConnection.getActiveShippings(requireContext()).subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(getShippingMethodResponseObserver())
    }

    private fun getShippingMethodResponseObserver(): Observer<in ShippingMethodResponse?> {
        return object : CustomObserver<ShippingMethodResponse?>(requireContext()) {
            override fun onSubscribe(d: Disposable) {
                super.onSubscribe(d)
                dialog.show()
            }

            override fun onNext(shippingMethodResponse: ShippingMethodResponse) {
                super.onNext(shippingMethodResponse)
                if (shippingMethodResponse.isAccessDenied) {
                    redirectToSignInSignUp()
                } else {
                    handleShippingMethodResponse(shippingMethodResponse)
                }
            }

            override fun onComplete() {
                super.onComplete()
                dialog.hide()
            }
        }
    }

    private fun setListeners() {
        binding.btnConfirmShippingMethod.setOnClickListener {
            if (!isShippingMethodSelected) {
                SnackbarHelper.getSnackbar(
                    requireActivity(), getString(R.string.checkout_select_a_shipping_method),
                    Snackbar.LENGTH_SHORT, SnackbarHelper.SnackbarType.TYPE_WARNING
                ).show()
            } else {
                updateShippingMethodInOrder(orderId, shippingMethod!!.id)
            }
        }
    }

    fun updateShippingMethodInOrder(orderId: Int, id: Int) {
        val updateOrderRequest = UpdateOrderRequest(UpdateOrderRequestModel(id, null))
        ApiConnection.updateOrderData(requireContext(), orderId, updateOrderRequest)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : CustomObserver<BaseResponse?>(requireActivity()) {
                override fun onSubscribe(d: Disposable) {
                    super.onSubscribe(d)
                    dialog.titleText = getString(R.string.updating_shipping_method_text)
                    dialog.show()
                }

                override fun onNext(response: BaseResponse) {
                    super.onNext(response)
                    handleUpdateResponse()
                }

                override fun onError(t: Throwable) {
                    super.onError(t)
                    SnackbarHelper.getSnackbar(
                        requireActivity(),
                        getString(R.string.error_something_went_wrong),
                        Snackbar.LENGTH_SHORT,
                        SnackbarHelper.SnackbarType.TYPE_WARNING
                    )
                }
            })
    }

    private fun handleUpdateResponse() {
        dialog.apply {
            hide()
            dismiss()
        }
        val intent = Intent()
        intent.putExtra(BundleConstant.BUNDLE_SELECTED_SHIPPING_METHOD, shippingMethod?.name)
        requireActivity().setResult(Activity.RESULT_OK, intent)
        requireActivity().finish()
    }


    private fun redirectToSignInSignUp() {
        AlertDialogHelper.showDefaultWarningDialogWithDismissListener(
            requireContext(),
            getString(R.string.error_login_failure),
            getString(R.string.access_denied_message)
        ) { sweetAlertDialog: SweetAlertDialog ->
            sweetAlertDialog.dismiss()
            AppSharedPref.clearCustomerData(requireContext())
            val signInSignUpIntent = Intent(requireContext(), SignInSignUpActivity::class.java)
            signInSignUpIntent.putExtra(
                BundleConstant.BUNDLE_KEY_CALLING_ACTIVITY,
                requireActivity().javaClass.getSimpleName()
            )
            requireContext().startActivity(signInSignUpIntent)
            requireActivity().finish()
        }
    }

    private fun handleShippingMethodResponse(shippingMethodResponse: ShippingMethodResponse) {
        binding.rvShippingMethod.adapter = ShippingMethodAdapter(
            requireContext(),
            shippingMethodResponse.shippingMethods,
            shippingMethodSelectionListener!!
        )
    }


    override fun updateShippingMethod(activeShippingMethod: ActiveShippingMethod) {
        this.shippingMethod = activeShippingMethod
        isShippingMethodSelected = true
    }

    override fun onDestroy() {
        super.onDestroy()
        if (dialog != null)
            dialog.hide()
    }
}